docs/csharp/versioning.md

```diff
@@ -0,0 +1,128 @@
+
+Here are some things to consider when trying to maintain backwards compatibility with older versions of your library:
+
+* Virtual methods: When you make a virtual method non-virtual in your new version it means that projects that override that method
+will have to be updated.
+* Method signatures: When updating a method behaviour requires you to change it's signature as well, you should instead create an override so that derived classes calling into that method will still work.
+You can always manipulate the old method signature to call into the new method signature so that implementation remains consistent.
+* Optional Method Arguments: When you make previously optional method arguments compulsory then all code that does not supply those arguments will need to be updated.
+On the other hand making compulsory arguments optional should have very little effect especially if it doesn't change the method's behaviour.
+* Obsolete attribute: You can use this attribute in your code to specify classes or class members that are deprecated and likely to be removed in future versions.
+This ensures developers utilizing your library are better prepared for breaking changes. More info on the `Obsolete` attribute [here](https://msdn.microsoft.com/en-us/library/22kk2b44(v=vs.90).aspx).
+
+## Consuming Libraries
+
+As a developer that consumes .NET libraries built by other developers you're most likely aware that a new version of a library might not be fully compatible with your project
+and you might often find yourself having to update your code to work with those changes.
+
+Lucky for you C# and the .NET ecosystem comes with features and techniques that allow us to easily update our app to work with new versions of libraries that might introduce breaking changes.
+
+### Assembly Binding Redirection
+
+You can use the `app.config` file to update the version of a library your app uses. By adding what is called a `binding redirect` your
+can use the new library version without having to recompile your app. The following example shows how you would update
+your app's `app.config` file to use the `1.0.1` patch version of `ReferencedLibrary` instead of the `1.0.0` version it was originally compiled with.
+
+```xml
+<dependentAssembly>
+    <assemblyIdentity name="ReferencedLibrary" publicKeyToken="32ab4ba45e0a69a1" culture="en-us" />
+    <bindingRedirect oldVersion="1.0.0" newVersion="1.0.1" />
+</dependentAssembly>
+```
+
+Go [here](https://msdn.microsoft.com/en-us/library/7wd6ex19(v=vs.110).aspx) to get more information on redirecting assembly versions.
+
+### new
+
+You use the `new` modifier to hide inherited members of a base class. This ensures resiliency of derived classes allowing you to update base classes the way you want in your library.
+
+Take the following example:
+
+```
+[!code-csharp[Sample usage of the 'new' modifier](/samples/csharp/versioning/new/Program.cs#sample)]
+```
+
+#### Output
```

Comment:
I wouldn't use a heading for this, since the text after the output does not belong under the heading.